apps/examples/grpc_route_client/route_guide_client.cc

```diff
@@ -252,40 +270,292 @@ class RouteGuideClient {
+      std::cout << "Visiting point "
+                << f.location().latitude()/kCoordFactor_ << ", "
+                << f.location().longitude()/kCoordFactor_ << std::endl;
+      writer->Write(f.location(), got_tag);
+      ret = cq.Next(&got_tag, &ok);
+      if (!ret || !ok || got_tag != (void*)1) {
+        std::cout << "Write Failed" << std::endl;
+        break;
+      }
+      std::this_thread::sleep_for(std::chrono::milliseconds(
+        delay_distribution(generator)));
+    }
+
+    writer->WritesDone(got_tag);
+    ret = cq.Next(&got_tag, &ok);
+    if (!ret || !ok || got_tag != (void*)1) {
+      std::cout << "WritesDone Failed" << std::endl;
+      return;
+    }
+
+    writer->Finish(&status, (void*)1);
+    ret = cq.Next(&got_tag, &ok);
+    if (!ret || !ok || got_tag != (void*)1) {
+      std::cout << "writer Finish Failed" << std::endl;
+      return;
+    }
+
+    if(status.ok()) {
+      std::cout << "Finished trip with " << stats.point_count() << " points\n"
+                << "Passed " << stats.feature_count() << " features\n"
+                << "Travelled " << stats.distance() << " meters\n"
+                << "It took " << stats.elapsed_time() << " seconds"
+                << std::endl;
+	  std::cout << "RecordRoute Done" << std::endl;
+    } else {
+      std::cout << "RecordRoute Failed." << std::endl;
+    }
+  }
+
+  void RouteChat() {
+    ClientContext context;
+    CompletionQueue cq;
+    Status status;
+
+    std::shared_ptr<ClientAsyncReaderWriter<RouteNote, RouteNote> > stream(
+      stub_->AsyncRouteChat(&context, &cq, (void *)1));
+
+    void* got_tag_read;
+    bool ok = false;
+    bool ret;
+
+    ret = cq.Next(&got_tag_read, &ok);
+    if (!ret || !ok || got_tag_read != (void*)1) {
+      std::cout << "RPC initialization Failed" << std::endl;
+      return;
+    }
+
+    std::thread writer([stream, &cq]() {
+      bool thd_ok = false;
+      bool thd_ret;
+      void* got_tag_write;
+
+      std::vector<RouteNote> notes{
+        MakeRouteNote("First message", 0, 0),
+        MakeRouteNote("Second message", 0, 1),
+        MakeRouteNote("Third message", 1, 0),
+        MakeRouteNote("Fourth message", 0, 0)};
+        for (const RouteNote& note : notes) {
+          std::cout << "Sending message " << note.message()
+                    << " at " << note.location().latitude() << ", "
+                    << note.location().longitude() << std::endl;
+          stream->Write(note, (void*)1);
+          thd_ret = cq.Next(&got_tag_write, &thd_ok);
+          if (!thd_ret || !thd_ok || got_tag_write != (void*)1) {
+            std::cout << "Write Failed" << std::endl;
+            break;
+          }
+        }
+        stream->WritesDone(got_tag_write);
+        thd_ret = cq.Next(&got_tag_write, &thd_ok);
+        if (!thd_ret || !thd_ok || got_tag_write != (void*)1) {
+          std::cout << "WriteDone Failed" << std::endl;
+          return;
```

Comment:
maybe not needed, as the thread was exiting anyway. Please ignore comment if not the case.